### ISSUE TITLE
how to backprop through the transformation matrix in tf.contrib.image.rotate

### ISSUE THREAD

**Issue by zhang-qiang-github** [voxelmorph/voxelmorph#70]
It's a very nice work. I am implementing a rigid registration method, and a problem stuck me. I use the CNN to predict the angle for rotation, and then calculate the loss between the target image and the wrap image. 

My test code is:

`
import tensorflow as tf

image_tensor = tf.placeholder(dtype=tf.float32, shape=[None,320,320,1])
target_tensor = tf.placeholder(dtype=tf.float32, shape=[None,320,320,1])

s = tf.concat([image_tensor, target_tensor],axis=3)
s = tf.layers.flatten(s)
w = tf.get_variable(initializer=tf.truncated_normal([204800,1], stddev=0.1),name='w')
b = tf.get_variable(initializer=tf.truncated_normal([1], stddev=0.1),name='b')
a = tf.matmul(s,w)+b
diff = tf.contrib.image.rotate(image_tensor, a[:,0], interpolation='BILINEAR') - target_tensor
loss = tf.reduce_sum(tf.square(diff))

optimizer = tf.train.GradientDescentOptimizer(0.001)
train = optimizer.minimize(loss)

`
However, there is an error: No gradients provided for any variable, check your graph for ops that do not support gradients

Is there no gradient of rotate function? Any suggestion is appreciated.

**Comment by adalca** [comment]
HI @zhang-qiang-github 

I honestly don't know about image.rotate (I'm a bit surprised if it doesn't have a gradient). 
However, you can use the `SpatialTransformer` in [neuron.layers](https://github.com/adalca/neuron/blob/master/neuron/layers.py) to apply affine transformation matrices -- so you could try using that, which we've had no problem with.

Hope this helps!

**Comment by zhang-qiang-github** [comment]
To be honest, I am not clear about how to use the SpatialTransformer. Could you please give me an example? For example, I have tensors for float_image and angle. How can I use the SpatialTransformer to get the wrap image?

**Comment by adalca** [comment]
`SpatialTransformer` is a Layer defined for keras models. If you have `float_image` and `angle` as keras tensors, you can do something like `warped_image_tensor = SpatialTransformer()([float_image, angle])`. This will give you the next layer tensor (output) in the model.

If they are just tensorflow tensors and you just want to do this one operation, it's a bit more complicated, but something *like* this would be the idea: first, set up the affine matrix from the rotation angle (see https://en.wikipedia.org/wiki/Affine_transformation#Image_Transformation Rotate row), then:

```
# get transform at each voxel from the affine matrix
trf = neuron.utils.affine_to_shift(affine, float_image.shape, shift_center=True)

# warp image according to transform
warped_image = neuron.utils.transform(transform, trf)
```

If you're planning to play with rigid registration, I recommend you learn more about affine transformation matrices anyway, and this can help you understand what's going on here.

**Comment by zhang-qiang-github** [comment]
How can I provide angle and translate (X, Y) seperately in SpatialTransformer?

**Comment by adalca** [comment]
given an angle and translation you can create an affine matrix. translations go in the top right entries of the affine matrix, `[0, 2]` and `[1, 2]` in numpy indexing.
